File: server/protocol.py
import json

VERSION = '0.0.1'

MESSAGE_SEPARATOR = b'\x00\x01\x00\x01\x00\x01'

MSG_DOMAIN_GAME = 'game'
MSG_DOMAIN_LOBBY = 'lobby'

MSG_DEFAULT = ''

# [SERVER] Base server messages
MSG_SRV_ERROR = 's.error'
MSG_SRV_HELLO = 's.hello'

# [SERVER] Queue message
MSG_SRV_QUEUE_STARTED = 's.q.started'
MSG_SRV_QUEUE_STOPPED = 's.q.stopped'
MSG_SRV_QUEUE_GAME_CREATED = 's.g.created'

# [SERVER] Game messages
MSG_SRV_GAME_BEGIN = 's.g.begin'
MSG_SRV_GAME_END = 's.g.end'
MSG_SRV_GAME_PLAYER_LEFT = 's.g.player_left'
MSG_SRV_GAME_TURN = 's.g.turn'
MSG_SRV_GAME_ACTION = 's.g.action'  # redirect player action
MSG_SRV_GAME_EFFECT = 's.g.effect'

# [CLIENT] Messages
MSG_CLI_QUEUE_START = 'c.q.start'
MSG_CLI_QUEUE_STOP = 'c.q.stop'
MSG_CLI_GAME_ACTION = 'c.g.action'

P_MSG_GAME_ID = 'game_id'

ENCODING = 'utf-8'


class Message(object):
    type = ''
    status = ''
    body = {}

    def __init__(self, domain, head, status='', **kwargs):
        self.head = head
        self.domain = domain
        self.status = status
        self.body = kwargs

    def __str__(self):
        return str(self.__dict__)


class GameMessage(Message):
    def __init__(self, head, game_id, *args, **kwargs):
        super().__init__(domain=MSG_DOMAIN_GAME, head=head, game_id=game_id, *args, **kwargs)


class LobbyMessage(Message):
    def __init__(self, head, *args, **kwargs):
        super().__init__(domain=MSG_DOMAIN_LOBBY, head=head, *args, **kwargs)


def serialize(obj):
    if isinstance(obj, Message):
        data = json.dumps(obj.__dict__)
    else:
        data = json.dumps(obj)
    return bytes(data, ENCODING)


def deserialize(data):
    if isinstance(data, bytes):
        data = data.decode(ENCODING)
    msg = Message(MSG_DOMAIN_GAME, MSG_DEFAULT)
    msg.__dict__ = json.loads(data)
    return msg

File: client/Assets/Scripts/Network/Protocol.cs
namespace Assets.Scripts.Network
{
    class Protocol
    {
        /// <summary>
        /// These values must be exact as server
        /// </summary>

        public const string MsgDomainGame = "game";
        public const string MsgDomainLobby = "lobby";

        public const string MsgDefault = "";

        // [SERVER] Base server messages
        public const string MsgSrvError = "s.error";

        public const string MsgSrvHello = "s.hello";

        // [SERVER] Queue message
        public const string MsgSrvQueueStarted = "s.q.started";

        public const string MsgSrvQueueStopped = "s.q.stopped";
        public const string MsgSrvQueueGameCreated = "s.g.created";

        // [SERVER] Game messages
        public const string MsgSrvGameBegin = "s.g.begin";

        public const string MsgSrvGameEnd = "s.g.end";
        public const string MsgSrvGamePlayerLeft = "s.g.player_left";
        public const string MsgSrvGameTurn = "s.g.turn";
        public const string MsgSrvGameAction = "s.g.action"; // redirect player action
        public const string MsgSrvGameEffect = "s.g.effect";

        // [CLIENT] Messages
        public const string MsgCliQueueStart = "c.q.start";

        public const string MsgCliQueueStop = "c.q.stop";
        public const string MsgCliGameAction = "c.g.action";

        public static byte[] MessageSeparator = new byte[] { 0x00, 0x01, 0x00, 0x01, 0x00, 0x01 };


        public const string KeyDomain = "domain";
        public const string KeyHead = "head";
        public const string KeyBody = "body";
        public const string KeyStatus = "status";

        public const string KeyShip = "ship";
        public const string KeyWeapon = "weapon";
        public const string KeyGameId = "game_id";
        public const string KeySide= "side";
    }
}